Age Gate

0x6273 opened the pull request:
# TODO
- [x] Integration tests need fixing. (Fixed in c9040dab4248f4f90b38aace72f443b29d688d37)
- [ ] The texts in the UI and ban message could probably be improved I think.
- [ ] Write integration tests for passing age gate and rules, and for getting banned by the age gate.
- [ ] Merge upstream into take2 and rebase this again once upstream updates its engine (needed for `DateOnly`)

## About the PR
Adds an age gate, requiring players to enter their date of birth before being let into the server.

## Why / Balance
To deter minors from playing on 18+ server. It won't stop all minors, but I think if it stops even a single one, it will be worth the slight inconvenience of people having to enter their DoB once before joining.

## Technical details
- Add a `PreJoinManager` that performs actions before a player has joined the game. It's similar to what's described in #390, but no prototypes and instead a cvar list of actions to perform in order (right now just `Rules` and `AgeGate`).
- Change `RulesManager` to make use of the `PreJoinManager`. It's no longer technically possible to join the game without having accepted the rules.
- Add the age gate, consisting of a manager, UI and UI controller, network messages, and a change to the `Player` DB model to add a boolean if the player has passed the age gate. 
- Fix a bug that caused an exception to be thrown on the server if the client disconnects just after connecting, before joining the game.

## Media
<img width="1870" height="1306" alt="image" src="https://github.com/user-attachments/assets/097127d3-c3a8-4221-9755-b1c66cdadff5" />


## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [x] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [x] I have added media to this PR or it does not require an ingame showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

0x6273 committed:
Gracefully handle a client disconnecting early

Fix an issue where if a client disconnects right after connecting, before
UserDbDataManager has had time to load the user data, it would throw an
exception.

0x6273 committed:
Add PreJoinManager

Add a manager that is responsible for making the player join the game
after connecting. It checks if there are any "prejoin actions" that need
to be performed first, such as presenting the player with a rules popup
or age gate.

0x6273 committed:
Add age gate

Add an age gate popup that requires the player to enter their date of
birth, checking that they are 18+ before letting them in.

This adds a boolean to the Player DB model, so running the EF Core
migrations script is required after applying the patch.

0x6273 commented:
Test fail isn't my fault, it's because pyyaml made a breaking change in a patch version...

It's fixed upstream https://github.com/space-wizards/RobustToolbox/compare/c8db7f98dbec...784a02c0e729


...actually the failing integration tests are my fault, gonna have to figure out what to do about those.

0x6273 committed:
Run EF Core migrations

moonheart08 reviewed (changes requested):
All of these new files need to go into _Citadel folders (with explicit MIT header) or your own folder (without it, if you want) if possible please.

0x6273 commented:
> All of these new files need to go into _Citadel folders (with explicit MIT header) or your own folder (without it, if you want) if possible please.

Yeah I think its own folder would be best. If every server that uses the age gate puts it in their specific folder it would make it difficult to share patches.

0x6273 committed:
Remove IPreJoinManager

0x6273 committed:
Change agegate UI text

0x6273 committed:
Move date picker to its own control

0x6273 committed:
Don't allow dates with year < 1000

Because someone could enter, say, "09" if they were born in the year
2009, which would actually be interpreted as the year 9.

0x6273 committed:
Move AgeGate to _Common folder

0x6273 committed:
Move PreJoinManager to _Common folder

0x6273 committed:
Fix namespaces

0x6273 commented:
> All of these new files need to go into _Citadel folders (with explicit MIT header) or your own folder (without it, if you want) if possible please.

I skipped the DatePicker since that should be upstreamed, but everything else has been moved to a `_Common` folder/namespace.

0x6273 committed:
DatePicker fixes

- Now uses `DateOnly` (currently fails sandbox, but PJB has promised to add
it in engine)
- Add MinDate and MaxDate fields
- Use constructor instead of parsing date string

0x6273 commented:
> The changes to upstream files should definitely carry `//Cit Change` or similar comments, so that any merge conflicts that may arise are clear.

Done. All changes to upstream files now has a `// Downstream change - needed for Age Gate` comment. 

> Also a minor nitpick (read: optional change) but the agegate UI could use a healthy dose of `HorizontalAlignment="Center"` on its elements; it'll help make it look a bit less jank

I'm not sure which elements you would want centered. But you should be able to push commits to this PR branch, or you can do it in a later PR.
